ingest/src/main/java/com/microsoft/azure/kusto/ingest/IngestClientBase.java

```diff
@@ -20,19 +26,34 @@ static boolean shouldCompress(CompressionType sourceCompressionType, IngestionPr
     }
 
     static String getIngestionEndpoint(String clusterUrl) {
-        if (clusterUrl.contains(INGEST_PREFIX)) {
+        if (clusterUrl == null || clusterUrl.contains(INGEST_PREFIX) || isReservedHostname(clusterUrl)) {
             return clusterUrl;
-        } else {
+        }
+        if (clusterUrl.contains(PROTOCOL_SUFFIX)) {
             return clusterUrl.replaceFirst(PROTOCOL_SUFFIX, PROTOCOL_SUFFIX + INGEST_PREFIX);
         }
+        return INGEST_PREFIX + clusterUrl;
     }
 
     static String getQueryEndpoint(String clusterUrl) {
-        if (clusterUrl.contains(INGEST_PREFIX)) {
-            return clusterUrl.replaceFirst(INGEST_PREFIX, "");
-        } else {
-            return clusterUrl;
+        return isReservedHostname(clusterUrl) ? clusterUrl : clusterUrl.replaceFirst(INGEST_PREFIX, "");
+    }
+
+    static boolean isReservedHostname(String rawUri) {
+        URI uri = URI.create(rawUri);
+        if (!uri.isAbsolute()) {
+            return true;
+        }
+        String authority = uri.getAuthority().toLowerCase();
+        boolean isIPFlag = InetAddressUtils.isIPv4Address(authority) || InetAddressUtils.isIPv6Address(authority);
+        boolean isLocalFlagg;
+        try {
```

Comment:
Using exceptions for controlflow is the only way to do it?
Also, wouldn't this actually send a DNS request to any of the addresses it gets? if so, this is bad.

Just check against the string "localhost", it's good enough.